core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import org.apache.iceberg.exceptions.ValidationException;
+
+public class SnapshotManager extends MergingSnapshotProducer<ManageSnapshots> implements ManageSnapshots {
+
+  private final TableOperations ops;
+  private SnapshotManagerOperation operation;
+  private TableMetadata base;
+  private Long targetSnapshotId = null;
+
+  enum SnapshotManagerOperation {
+    CHERRYPICK,
+    ROLLBACK
+  }
+
+  SnapshotManager(TableOperations ops) {
+    super(ops);
+    this.ops = ops;
+    this.base = ops.current();
+  }
+
+  @Override
+  protected ManageSnapshots self() {
+    return this;
+  }
+
+  @Override
+  protected String operation() {
+    Snapshot manageSnapshot = base.snapshot(targetSnapshotId);
+    return manageSnapshot.operation();
+  }
+
+  @Override
+  public ManageSnapshots cherrypick(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.CHERRYPICK;
```

Comment:
Why set operation here and then use matching control flow in `apply`? Seems like it would be easier to pick the commit right here instead. Then keep track of the operation that was applied to be able to fail immediately in `setSnapshotTo` or `rollbackToTime`. A bonus advantage is that the stack trace will show the problem in the configuration methods that are called, rather than in `apply`.

Reply:
Got it, I'v moved all the cherrypicking of appends to `SnapshotManager.cherrypick()` method.

Although I still need to be able to differentiate the operation in `SnapshotManager.apply()` because if it's _CHERRYPICK_ operation then we call `super.apply()` to create the new snapshot using SnapshotProducer's impl , and if it's _ROLLBACK_ we just return the `base.snapshot(targetSnapshotId)`